File: backend/utils/gpsEncryption.js
const crypto = require('crypto');
require('dotenv').config();

/**
 * GPS Location Encryption Utility
 * Provides AES-256-CBC encryption/decryption for sensitive GPS coordinates
 */
class GPSEncryption {
    constructor() {
        // Get encryption key and IV from environment
        this.encryptionKey = process.env.GPS_ENCRYPTION_KEY || 'MargSetu2024SecureGPSLocationKey32B';
        this.encryptionIV = process.env.GPS_ENCRYPTION_IV || 'MargSetuGPSIV16B';
        this.encryptionEnabled = process.env.GPS_ENCRYPTION_ENABLED === 'true';
        
        // Ensure key is exactly 32 bytes for AES-256
        if (this.encryptionKey.length !== 32) {
            this.encryptionKey = this.encryptionKey.padEnd(32, '0').substring(0, 32);
        }
        
        // Ensure IV is exactly 16 bytes
        if (this.encryptionIV.length !== 16) {
            this.encryptionIV = this.encryptionIV.padEnd(16, '0').substring(0, 16);
        }
        
        console.log(`📡 GPS Encryption: ${this.encryptionEnabled ? 'ENABLED' : 'DISABLED'}`);
    }
    
    /**
     * Encrypt GPS coordinates
     * @param {Object} gpsData - GPS data object {busId, latitude, longitude, timestamp}
     * @returns {String} - Encrypted GPS data as base64 string
     */
    encryptGPS(gpsData) {
        if (!this.encryptionEnabled) {
            return JSON.stringify(gpsData);
        }
        
        try {
            const plaintext = JSON.stringify(gpsData);
            // Use modern crypto API with explicit key and IV
            const key = Buffer.from(this.encryptionKey, 'utf8');
            const iv = Buffer.from(this.encryptionIV, 'utf8');
            
            const cipher = crypto.createCipheriv('aes-256-cbc', key, iv);
            
            let encrypted = cipher.update(plaintext, 'utf8', 'base64');
            encrypted += cipher.final('base64');
            
            console.log(`🔒 GPS data encrypted for ${gpsData.busId}`);
            return encrypted;
            
        } catch (error) {
            console.error('❌ GPS encryption failed:', error);
            // Fallback to unencrypted if encryption fails
            return JSON.stringify(gpsData);
        }
    }
    
    /**
     * Decrypt GPS coordinates
     * @param {String} encryptedData - Encrypted GPS data as base64 string
     * @returns {Object} - Decrypted GPS data object
     */
    decryptGPS(encryptedData) {
        if (!this.encryptionEnabled) {
            try {
                return JSON.parse(encryptedData);
            } catch (e) {
                console.error('❌ Failed to parse unencrypted GPS data:', e);
                return null;
            }
        }
        
        try {
            // Use modern crypto API with explicit key and IV
            const key = Buffer.from(this.encryptionKey, 'utf8');
            const iv = Buffer.from(this.encryptionIV, 'utf8');
            
            const decipher = crypto.createDecipheriv('aes-256-cbc', key, iv);
            
            let decrypted = decipher.update(encryptedData, 'base64', 'utf8');
            decrypted += decipher.final('utf8');
            
            const gpsData = JSON.parse(decrypted);
            console.log(`🔓 GPS data decrypted for ${gpsData.busId || 'unknown'}`);
            
            return gpsData;
            
        } catch (error) {
            console.error('❌ GPS decryption failed:', error);
            console.error('❌ Encrypted data:', encryptedData.substring(0, 50) + '...');
            
            // Try to parse as unencrypted fallback
            try {
                return JSON.parse(encryptedData);
            } catch (e) {
                console.error('❌ Fallback parsing also failed');
                return null;
            }
        }
    }
    
    /**
     * Encrypt GPS coordinates with additional metadata
     * @param {String} busId - Bus identifier
     * @param {Number} latitude - GPS latitude
     * @param {Number} longitude - GPS longitude
     * @param {Object} metadata - Additional metadata (timestamp, speed, etc.)
     * @returns {String} - Encrypted GPS package
     */
    encryptGPSLocation(busId, latitude, longitude, metadata = {}) {
        const gpsPackage = {
            busId: busId,
            latitude: parseFloat(latitude),
            longitude: parseFloat(longitude),
            timestamp: metadata.timestamp || Date.now(),
            speed: metadata.speed || 0,
            accuracy: metadata.accuracy || 0,
            source: metadata.source || 'SMS',
            version: '1.0'
        };
        
        return this.encryptGPS(gpsPackage);
    }
    
    /**
     * Validate decrypted GPS data
     * @param {Object} gpsData - Decrypted GPS data
     * @returns {Boolean} - Whether data is valid
     */
    validateGPSData(gpsData) {
        if (!gpsData || typeof gpsData !== 'object') {
            return false;
        }
        
        const { busId, latitude, longitude } = gpsData;
        
        // Check required fields
        if (!busId || latitude === undefined || longitude === undefined) {
            return false;
        }
        
        // Validate coordinate ranges
        if (typeof latitude !== 'number' || typeof longitude !== 'number') {
            return false;
        }
        
        if (latitude < -90 || latitude > 90) {
            return false;
        }
        
        if (longitude < -180 || longitude > 180) {
            return false;
        }
        
        return true;
    }
    
    /**
     * Create encrypted GPS message for SMS
     * @param {String} busId - Bus identifier
     * @param {Number} latitude - GPS latitude
     * @param {Number} longitude - GPS longitude
     * @returns {String} - SMS-ready encrypted GPS message
     */
    createEncryptedSMSMessage(busId, latitude, longitude) {
        const encryptedData = this.encryptGPSLocation(busId, latitude, longitude, {
            timestamp: Date.now(),
            source: 'SMS_GATEWAY'
        });
        
        // Format: GPS_ENC:encrypted_data
        return `GPS_ENC:${encryptedData}`;
    }
    
    /**
     * Parse encrypted GPS message from SMS
     * @param {String} smsMessage - SMS message content
     * @returns {Object|null} - Parsed GPS data or null if invalid
     */
    parseEncryptedSMSMessage(smsMessage) {
        // Check for encrypted GPS format
        if (smsMessage.startsWith('GPS_ENC:')) {
            const encryptedData = smsMessage.substring(8); // Remove "GPS_ENC:" prefix
            const decryptedData = this.decryptGPS(encryptedData);
            
            if (this.validateGPSData(decryptedData)) {
                return decryptedData;
            }
        }
        
        // Fallback: try to parse as plain GPS format
        const plainGPSMatch = smsMessage.match(/^GPS:([A-Z0-9]+),(-?\d+\.?\d*),(-?\d+\.?\d*)/i);
        if (plainGPSMatch) {
            const [, busId, lat, lng] = plainGPSMatch;
            return {
                busId: busId,
                latitude: parseFloat(lat),
                longitude: parseFloat(lng),
                timestamp: Date.now(),
                source: 'SMS_PLAIN',
                version: '1.0'
            };
        }
        
        return null;
    }
}

// Export singleton instance
const gpsEncryption = new GPSEncryption();
module.exports = gpsEncryption;

// Export class for testing
module.exports.GPSEncryption = GPSEncryption;

File: driver-app/app/src/main/java/com/margsetu/driver/utils/GPSEncryption.kt
package com.margsetu.driver.utils

import android.util.Base64
import android.util.Log
import org.json.JSONObject
import javax.crypto.Cipher
import javax.crypto.spec.SecretKeySpec
import javax.crypto.spec.IvParameterSpec

/**
 * GPS Encryption Utility for Driver App
 * Provides AES-256-ECB encryption for GPS coordinates before transmission
 */
class GPSEncryption {
    
    companion object {
        private const val TAG = "GPSEncryption"
        private const val ALGORITHM = "AES"
        private const val TRANSFORMATION = "AES/CBC/PKCS5Padding"
        
        // Default encryption key and IV (should match backend)
        private const val DEFAULT_ENCRYPTION_KEY = "MargSetu2024SecureGPSLocationKey32B"
        private const val DEFAULT_ENCRYPTION_IV = "MargSetuGPSIV16B"
        
        @Volatile
        private var INSTANCE: GPSEncryption? = null
        
        fun getInstance(): GPSEncryption {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: GPSEncryption().also { INSTANCE = it }
            }
        }
    }
    
    private val encryptionKey: String = DEFAULT_ENCRYPTION_KEY
    private val encryptionIV: String = DEFAULT_ENCRYPTION_IV
    private var isEncryptionEnabled: Boolean = true
    
    init {
        Log.d(TAG, "GPS Encryption initialized: ${if (isEncryptionEnabled) "ENABLED" else "DISABLED"}")
    }
    
    /**
     * Enable or disable GPS encryption
     */
    fun setEncryptionEnabled(enabled: Boolean) {
        isEncryptionEnabled = enabled
        Log.d(TAG, "GPS Encryption: ${if (enabled) "ENABLED" else "DISABLED"}")
    }
    
    /**
     * Encrypt GPS coordinates
     * @param busNumber Bus identifier
     * @param latitude GPS latitude
     * @param longitude GPS longitude
     * @param metadata Additional metadata
     * @return Encrypted GPS data as base64 string
     */
    fun encryptGPSLocation(
        busNumber: String,
        latitude: Double,
        longitude: Double,
        metadata: Map<String, Any> = emptyMap()
    ): String {
        if (!isEncryptionEnabled) {
            return createPlainGPSMessage(busNumber, latitude, longitude)
        }
        
        try {
            val gpsData = JSONObject().apply {
                put("busId", busNumber)
                put("latitude", latitude)
                put("longitude", longitude)
                put("timestamp", System.currentTimeMillis())
                put("speed", metadata["speed"] ?: 0)
                put("accuracy", metadata["accuracy"] ?: 0)
                put("source", metadata["source"] ?: "DRIVER_APP")
                put("version", "1.0")
            }
            
            val plaintext = gpsData.toString()
            val encryptedData = encrypt(plaintext)
            
            Log.d(TAG, "GPS data encrypted for $busNumber")
            return encryptedData
            
        } catch (e: Exception) {
            Log.e(TAG, "GPS encryption failed", e)
            // Fallback to plain message
            return createPlainGPSMessage(busNumber, latitude, longitude)
        }
    }
    
    /**
     * Create encrypted SMS message for GPS
     * @param busNumber Bus identifier
     * @param latitude GPS latitude
     * @param longitude GPS longitude
     * @return SMS-ready encrypted GPS message
     */
    fun createEncryptedSMSMessage(busNumber: String, latitude: Double, longitude: Double): String {
        val encryptedData = encryptGPSLocation(
            busNumber = busNumber,
            latitude = latitude,
            longitude = longitude,
            metadata = mapOf(
                "timestamp" to System.currentTimeMillis(),
                "source" to "DRIVER_APP_SMS"
            )
        )
        
        return if (isEncryptionEnabled) {
            "GPS_ENC:$encryptedData"
        } else {
            encryptedData // Already formatted as plain GPS message
        }
    }
    
    /**
     * Create encrypted HTTP payload for GPS
     * @param busNumber Bus identifier
     * @param latitude GPS latitude
     * @param longitude GPS longitude
     * @param driverName Driver name
     * @return Encrypted GPS data for HTTP transmission
     */
    fun createEncryptedHTTPPayload(
        busNumber: String, 
        latitude: Double, 
        longitude: Double,
        driverName: String = "Unknown"
    ): String {
        return encryptGPSLocation(
            busNumber = busNumber,
            latitude = latitude,
            longitude = longitude,
            metadata = mapOf(
                "timestamp" to System.currentTimeMillis(),
                "source" to "DRIVER_APP_HTTP",
                "driverName" to driverName
            )
        )
    }
    
    /**
     * Encrypt string data using AES-CBC
     */
    private fun encrypt(plaintext: String): String {
        try {
            val key = prepareKey(encryptionKey)
            val iv = prepareIV(encryptionIV)
            val cipher = Cipher.getInstance(TRANSFORMATION)
            cipher.init(Cipher.ENCRYPT_MODE, key, iv)
            
            val encryptedBytes = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))
            return Base64.encodeToString(encryptedBytes, Base64.NO_WRAP)
            
        } catch (e: Exception) {
            Log.e(TAG, "Encryption failed", e)
            throw e
        }
    }
    
    /**
     * Prepare encryption key to correct length (same as Node.js backend)
     */
    private fun prepareKey(key: String): SecretKeySpec {
        // Use same logic as Node.js: pad with '0' character, not null bytes
        val keyStr = if (key.length != 32) {
            key.padEnd(32, '0').substring(0, 32)
        } else {
            key
        }
        
        val keyBytes = keyStr.toByteArray(Charsets.UTF_8)
        return SecretKeySpec(keyBytes, ALGORITHM)
    }
    
    /**
     * Prepare IV to correct length (same as Node.js backend)
     */
    private fun prepareIV(iv: String): IvParameterSpec {
        // Use same logic as Node.js: pad with '0' character, not null bytes
        val ivStr = if (iv.length != 16) {
            iv.padEnd(16, '0').substring(0, 16)
        } else {
            iv
        }
        
        val ivBytes = ivStr.toByteArray(Charsets.UTF_8)
        return IvParameterSpec(ivBytes)
    }
    
    /**
     * Create plain GPS message (fallback)
     */
    private fun createPlainGPSMessage(busNumber: String, latitude: Double, longitude: Double): String {
        return "GPS:$busNumber,$latitude,$longitude,0.0,0,${System.currentTimeMillis()}"
    }
    
    /**
     * Validate GPS coordinates
     */
    fun validateCoordinates(latitude: Double, longitude: Double): Boolean {
        return latitude >= -90.0 && latitude <= 90.0 && 
               longitude >= -180.0 && longitude <= 180.0
    }
    
    /**
     * Test encryption/decryption functionality
     */
    fun testEncryption(): Boolean {
        return try {
            val testData = "Test GPS data for encryption"
            val encrypted = encrypt(testData)
            
            val success = encrypted.isNotEmpty()
            Log.d(TAG, "Encryption test: ${if (success) "PASSED" else "FAILED"}")
            success
            
        } catch (e: Exception) {
            Log.e(TAG, "Encryption test failed", e)
            false
        }
    }
}